plugins/package-configuration-providers/dos/src/main/kotlin/DosPackageConfigurationProvider.kt

```diff
@@ -0,0 +1,164 @@
+
+    override fun parseConfig(options: Options, secrets: Options) =
+        DosPackageConfigurationProviderConfig(
+            url = options.getValue("url"),
+            timeout = options["timeout"]?.toLongOrNull(),
+            token = secrets.getValue("token")
+        )
+}
+
+/**
+ * A [PackageConfigurationProvider] that loads [PackageConfiguration]s from a Double Open Server instance.
+ */
+class DosPackageConfigurationProvider(config: DosPackageConfigurationProviderConfig) : PackageConfigurationProvider {
+    private val service = DosService.create(
+        config.url,
+        config.token,
+        config.timeout?.let { Duration.ofSeconds(it) }
+    )
+
+    private val client = DosClient(service)
+
+    override fun getPackageConfigurations(packageId: Identifier, provenance: Provenance): List<PackageConfiguration> {
+        val purl = packageId.toPurl(provenance.toPurlExtras())
+
+        val packageResults = runBlocking { client.getPackageConfiguration(purl) }
+        if (packageResults == null) {
+            logger.info { "Unable to find package configuration for '$purl'." }
+            return emptyList()
+        }
+
+        return if (packageResults.licenseConclusions.isEmpty() && packageResults.pathExclusions.isEmpty()) {
+            emptyList()
+        } else {
+            val packageConfiguration = generatePackageConfiguration(
+                id = packageId,
+                provenance = provenance,
+                packageResults = packageResults
+            )
+
+            logger.info { "Found license conclusions for $purl: ${packageResults.licenseConclusions}" }
+            logger.info { "Found path exclusions for $purl: ${packageResults.pathExclusions}" }
+
+            listOf(packageConfiguration)
+        }
+    }
+}
+
+internal fun generatePackageConfiguration(
+    id: Identifier,
+    provenance: Provenance,
+    packageResults: PackageConfigurationResponseBody
+): PackageConfiguration {
+    val sourceArtifactUrl = if (provenance is ArtifactProvenance) {
+        provenance.sourceArtifact.url
+    } else {
+        null
+    }
+
+    val vcs = if (provenance is RepositoryProvenance) {
+        VcsMatcher(
+            type = provenance.vcsInfo.type,
+            url = provenance.vcsInfo.url,
+            revision = provenance.resolvedRevision
+        )
+    } else {
+        null
+    }
+
+    val licenseFindingCurations = packageResults.licenseConclusions.map { licenseConclusion ->
+        val detected = licenseConclusion.detectedLicenseExpressionSPDX?.takeUnless { it.isEmpty() }
+            ?.let { SpdxExpression.parse(it) }
+
+        check(licenseConclusion.concludedLicenseExpressionSPDX.isNotEmpty())
+        val concluded = SpdxExpression.parse(licenseConclusion.concludedLicenseExpressionSPDX)
+
+        LicenseFindingCuration(
+            path = licenseConclusion.path,
+            detectedLicense = detected?.let { associateLicensesWithExceptions(it) },
```

Comment:
I don't understand what / why associate with exception is done here.

Reply:
The DOS scanner uses ScanCode on the server-side, so it suffers from the same shortcomings in ScanCode's results as when running locally. This means we also need to fixup stand-alone exception findings here.

Reply:
Hm, but these are license finding curations, not license findings. 
Shouldn't the curations target the actual detected licenses (which has associated exceptions)? (Instead of matching against the unpatched detected licenses which the curator normally does not know about)

Reply:
> Shouldn't the curations target the actual detected licenses (which has associated exceptions)?

Exactly that's the reason for calling `associateLicensesWithExceptions()` also here: Remember that this is the *matcher* for the detected license that ORT already "knows", and that known detected license was post-processed with `associateLicensesWithExceptions()` previously. So in order to match that post-processed detected license, we need to apply the same post-processing here as well.

Reply:
Just curious, the license finding curations in ORT we write against the "post processed" findings. In particular, we do not run `associateLicensesWithExceptions()`. Why is this different for DOS and why don't these curations get fixed on the backend / in the database ?

Reply:
> Just curious, the license finding curations in ORT we write against the "post processed" findings.

I agree with this, but...

> In particular, we do not run `associateLicensesWithExceptions()`.

... this is IMO a contradiction to that. Post-processing includes running `associateLicensesWithExceptions()`, and we do that at

https://github.com/oss-review-toolkit/ort/blob/f995050ff24808ffc95d77772963ca19d15f8141/plugins/scanners/scancode/src/main/kotlin/ScanCodeResultModelMapper.kt#L154

So any detected license returned by ScanCode got post-processed by `associateLicensesWithExceptions()`. Hence also everything that we match against this license needs to be be post-processed by `associateLicensesWithExceptions()`.

Reply:
> So any detected license returned by ScanCode got post-processed by `associateLicensesWithExceptions()`. Hence also everything that we match against this license needs to be be post-processed by `associateLicensesWithExceptions()`.

This bit I still do not understand. What I do understand is that in ORT we call the postprocessing in the scanner, but not for the license finding curations. Because the license finding curation are written against the post processed licenses, not against the raw ones. In DOS this aspect (for the license finding curations, not scanner) seems to be different. Why?

Reply:
I'm still not 100% certain about your question, but I guess it can be rephrased as "Why does DOS here need to call `associateLicensesWithExceptions()`, but the other `PackageConfigurationProvider`s don't?".

The answer to that question is bit difficult to give without knowing the DOS scanner implementation (which I probably should have contributed first, but I wanted to do the "easier" `PackageConfigurationProvider` contribution first, as it's less code):

The DOS scanner backend is written in TypeScript, not in Koltin, and thus cannot make use of any ORT functionality (that was a design decision that probably needs to be revisited in the context of ORT server, but that's another topic). This means that `associateLicensesWithExceptions()` is not available to the DOS scanner, so it can't post-process the ScanCode results with it. So we "postpone" the post-processing that would usually be done by the `ScanCodeResultModelMapper`, which the DOS scanner cannot use, to here.

Reply:
This comment clarifies what I've been after. If this information is added as a code comment, I'd be ok with keeping the `associate*` call.

Reply:
Thanks!